tiledb/api/src/context.rs

```diff
@@ -68,95 +68,110 @@ impl Context {
         let res = unsafe { ffi::tiledb_ctx_alloc(cfg.capi(), &mut c_ctx) };
         if res == ffi::TILEDB_OK {
             Ok(Context {
-                raw: RawContext::Owned(c_ctx),
+                raw: Arc::new(Mutex::new(RawContext { raw: c_ctx })),
             })
         } else {
             Err(Error::LibTileDB(String::from("Could not create context")))
         }
     }
 
-    pub fn get_stats(&self) -> TileDBResult<String> {
-        let mut c_json = std::ptr::null_mut::<std::os::raw::c_char>();
-        let res = unsafe {
-            ffi::tiledb_ctx_get_stats(
-                *self.raw,
-                &mut c_json as *mut *mut std::os::raw::c_char,
-            )
+    pub fn capi_call<Callable>(&self, action: Callable) -> TileDBResult<()>
+    where
+        Callable: FnOnce(*mut ffi::tiledb_ctx_t) -> i32,
+    {
+        // Scope the lock so we can get the error if one happens.
+        let res = {
+            // Docs say that the error result from a poisoned lock are usually
+            // just propagated to all threads by just calling unwrap on locks.
+            let raw_ctx = self.raw.lock().unwrap();
```

Comment:
Per my above comment I think this should be reverted, but...

It is also done incorrectly as the lock must be held before `action` begins and released after `expect_last_error` ends.

Reply:
Ohhhhh, good catch!

Reply:
Fixed.